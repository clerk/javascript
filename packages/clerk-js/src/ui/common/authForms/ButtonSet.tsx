import { ButtonWithSpinner } from '@clerk/shared/components/button';
import React from 'react';
import { svgUrl } from 'ui/common/constants';

type Flow = 'sign-in' | 'sign-up';

export interface ButtonSetOptions {
  id: string;
  name: string;
  strategy: string;
}

export type ButtonSetProps<T> = {
  buttonClassName?: string;
  className?: string;
  error?: string;
  flow: Flow;
  handleClick: (e: React.MouseEvent, provider: T) => void;
  options: ButtonSetOptions[];
};

export function ButtonSet<T>({
  buttonClassName,
  className,
  error,
  flow,
  handleClick,
  options,
}: ButtonSetProps<T>): JSX.Element | null {
  const [isLoading, setIsLoading] = React.useState(() =>
    options.map(() => false),
  );
  const alreadyLoading = isLoading.some(o => o);

  React.useEffect(() => {
    if (!alreadyLoading) {
      return;
    }
    setIsLoading(options.map(() => false));
  }, [alreadyLoading, error, options]);

  const handleOptionClick = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    i: number,
  ) => {
    handleClick(e, (options[i].strategy as unknown) as T);
    setIsLoading(() => {
      const newState = [...isLoading];
      newState[i] = true;
      return newState;
    });
  };

  if (options.length <= 0) {
    return null;
  }

  return (
    <div className={className}>
      {options.map(({ id, name }, index: number) => {
        return (
          <ButtonWithSpinner
            key={`button-set-${id}-${index}`}
            isLoading={isLoading[index]}
            disabled={alreadyLoading}
            onClick={e => handleOptionClick(e, index)}
            flavor='outline'
            className={buttonClassName}
          >
            <img alt={name} src={svgUrl(id)} />
            <span>{getButtonLabel(flow, name)}</span>
          </ButtonWithSpinner>
        );
      })}
    </div>
  );
}

function getButtonLabel(flow: Flow, provider: string) {
  switch (flow) {
    case 'sign-up':
      return `Sign up with ${provider}`;
    case 'sign-in':
    default:
      return `Sign in with ${provider}`;
  }
}
