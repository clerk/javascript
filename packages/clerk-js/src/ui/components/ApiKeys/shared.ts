import { useClerk } from '@clerk/shared/react';
import { useState } from 'react';

import { useFetch } from '../../hooks';
import type { Expiration } from './CreateApiKeyForm';

function getTimeLeftInSeconds(expirationOption: Expiration) {
  if (expirationOption === 'never') {
    return;
  }
  const now = new Date();
  const future = new Date(now);
  if (expirationOption === '30d') {
    future.setDate(future.getDate() + 30);
  } else if (expirationOption === '90d') {
    future.setDate(future.getDate() + 90);
  } else {
    throw new Error('TODO: Improve time handling');
  }
  const diffInMs = future.getTime() - now.getTime();
  const diffInSecs = Math.floor(diffInMs / 1000);
  return diffInSecs;
}

export function useApiKeys({ subject, perPage = 5 }: { subject: string; perPage?: number }) {
  const clerk = useClerk();
  const {
    data: apiKeys,
    isLoading,
    revalidate,
  } = useFetch(clerk.getApiKeys, { subject }, undefined, `api-key-source-${subject}`);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState('');
  const itemsPerPage = perPage;

  const filteredApiKeys = (apiKeys ?? []).filter(key => key.name.toLowerCase().includes(search.toLowerCase()));
  const itemCount = filteredApiKeys.length;
  const pageCount = Math.max(1, Math.ceil(itemCount / itemsPerPage));
  const startingRow = itemCount > 0 ? (page - 1) * itemsPerPage + 1 : 0;
  const endingRow = Math.min(page * itemsPerPage, itemCount);
  const paginatedApiKeys = filteredApiKeys.slice(startingRow - 1, endingRow);

  const handleCreate = async (params: {
    name: string;
    description?: string;
    expiration: Expiration;
    closeFn: () => void;
  }) => {
    await clerk.createApiKey({
      name: params.name,
      creationReason: params.description,
      secondsUntilExpiration: getTimeLeftInSeconds(params.expiration),
    });
    params.closeFn();
    revalidate();
  };

  const revokeApiKey = async (apiKeyID: string) => {
    await clerk.revokeApiKey({ apiKeyID, revocationReason: 'Revoked by user' });
    setPage(1);
    revalidate();
  };

  return {
    apiKeys: paginatedApiKeys,
    isLoading: isLoading ?? false,
    revokeApiKey,
    search,
    setSearch,
    page,
    setPage,
    pageCount,
    itemCount,
    startingRow,
    endingRow,
    handleCreate,
  };
}
