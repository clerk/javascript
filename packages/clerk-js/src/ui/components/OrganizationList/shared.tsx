import type { UserOrganizationInvitationResource } from '@clerk/types';
import type { PropsWithChildren } from 'react';
import { forwardRef } from 'react';

import { Box, Button, Col, descriptors, Flex, Spinner } from '../../customizables';
import { OrganizationPreview, PreviewButton } from '../../elements';
import { ArrowRightIcon } from '../../icons';
import type { ThemableCssProp } from '../../styledSystem';
import { common } from '../../styledSystem';

export const PreviewListItems = (props: PropsWithChildren) => {
  return (
    <Col
      elementDescriptor={descriptors.organizationListPreviewItems}
      sx={t => ({
        maxHeight: `calc(8 * ${t.sizes.$12})`,
        overflowY: 'auto',
        ...common.unstyledScrollbar(t),
      })}
    >
      {props.children}
    </Col>
  );
};

const sharedStyles: ThemableCssProp = t => ({
  height: t.space.$14,
  padding: `${t.space.$2} ${t.space.$8}`,
});

export const PreviewListItem = (
  props: PropsWithChildren<{
    organizationData: UserOrganizationInvitationResource['publicOrganizationData'];
  }>,
) => {
  return (
    <Flex
      align='center'
      gap={2}
      sx={[
        _ => ({
          minHeight: 'unset',
          justifyContent: 'space-between',
        }),
        sharedStyles,
      ]}
      elementDescriptor={descriptors.organizationListPreviewItem}
    >
      <OrganizationPreview
        elementId='organizationList'
        avatarSx={t => ({ width: t.sizes.$10, height: t.sizes.$10 })}
        mainIdentifierSx={t => ({
          fontSize: t.fontSizes.$xl,
          color: t.colors.$colorText,
        })}
        organization={props.organizationData}
      />
      {props.children}
    </Flex>
  );
};

export const PreviewListSpinner = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <Box
      ref={ref}
      sx={t => ({
        width: '100%',
        height: t.space.$12,
        position: 'relative',
      })}
    >
      <Box
        sx={{
          margin: 'auto',
          position: 'absolute',
          left: '50%',
          top: '50%',
          transform: 'translateY(-50%) translateX(-50%)',
        }}
      >
        <Spinner
          size='md'
          colorScheme='primary'
        />
      </Box>
    </Box>
  );
});

export const PreviewListItemButton = (props: Parameters<typeof Button>[0]) => {
  return (
    <Button
      elementDescriptor={descriptors.organizationListPreviewItemActionButton}
      textVariant='buttonExtraSmallBold'
      variant='outline'
      colorScheme='neutral'
      size='sm'
      {...props}
    />
  );
};

export const OrganizationListPreviewButton = (props: PropsWithChildren<{ onClick: () => void | Promise<void> }>) => {
  return (
    <PreviewButton
      elementDescriptor={descriptors.organizationListPreviewButton}
      sx={[sharedStyles]}
      icon={ArrowRightIcon}
      iconProps={{
        size: 'lg',
      }}
      showIconOnHover={false}
      {...props}
    />
  );
};
