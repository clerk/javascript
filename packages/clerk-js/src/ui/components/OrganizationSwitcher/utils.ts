import type { ClerkPaginatedResponse } from '@clerk/types';

import type { useCoreOrganizationList } from '../../contexts';

export const organizationListParams = {
  userMemberships: {
    infinite: true,
  },
  userInvitations: {
    infinite: true,
  },
  userSuggestions: {
    infinite: true,
    status: ['pending', 'accepted'],
  },
} satisfies Parameters<typeof useCoreOrganizationList>[0];

export const updateCacheInPlace =
  (userSuggestions: any) =>
  <T extends { id: string }>(result: T): any => {
    userSuggestions?.unstable__mutate?.(result, {
      populateCache: (updatedItem: T, itemsInfinitePages: ClerkPaginatedResponse<T>[]) => {
        return itemsInfinitePages.map(item => {
          const newData = item.data.map(obj => {
            if (obj.id === updatedItem.id) {
              return {
                ...updatedItem,
              };
            }

            return obj;
          });
          return { ...item, data: newData };
        });
      },
      // Since `accept` gives back the updated information,
      // we don't need to revalidate here.
      revalidate: false,
    });
  };

export const removeItemFromPaginatedCache =
  (userInvitations: any) =>
  <T extends { id: string }>(result: T): any => {
    userInvitations?.unstable__mutate?.(result, {
      populateCache: (updatedItem: T, itemsInfinitePages: ClerkPaginatedResponse<T>[]) => {
        const prevTotalCount = itemsInfinitePages[itemsInfinitePages.length - 1].total_count;

        return itemsInfinitePages.map(item => {
          const newData = item.data.filter(obj => {
            return obj.id !== updatedItem.id;
          });
          return { ...item, data: newData, total_count: prevTotalCount - 1 };
        });
      },
      // Since `accept` gives back the updated information,
      // we don't need to revalidate here.
      revalidate: false,
    });
  };
