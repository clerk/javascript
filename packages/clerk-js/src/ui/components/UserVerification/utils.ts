import type { SignInFactor } from '@clerk/types';

export function sortByPrimaryFactor(a: SignInFactor, b: SignInFactor) {
  if ('primary' in a && a.primary && !('primary' in b && b.primary)) {
    return -1;
  }

  if ('primary' in b && b.primary && !('primary' in a && a.primary)) {
    return 1;
  }

  return 0;
}
