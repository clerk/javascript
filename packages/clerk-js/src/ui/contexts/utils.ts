import type { Clerk } from '@clerk/types';

import { clerkCoreErrorContextProviderNotFound, clerkCoreErrorNoClerkSingleton } from '../../core/errors';

export function assertClerkSingletonExists(clerk: Clerk | undefined): asserts clerk is Clerk {
  if (!clerk) {
    clerkCoreErrorNoClerkSingleton();
  }
}

export function assertContextExists(contextVal: unknown, providerName: string): asserts contextVal {
  if (!contextVal) {
    clerkCoreErrorContextProviderNotFound(providerName);
  }
}
