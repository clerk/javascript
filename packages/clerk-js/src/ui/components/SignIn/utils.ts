import { titleize } from '@clerk/shared/underscore';
import { isWebAuthnSupported } from '@clerk/shared/webauthn';
import type { PreferredSignInStrategy, SignInFactor, SignInResource, SignInStrategy } from '@clerk/types';
import type { FormControlState } from 'ui/utils';

import { PREFERRED_SIGN_IN_STRATEGIES } from '../../common/constants';
import { otpPrefFactorComparator, passwordPrefFactorComparator } from '../../utils/factorSorting';

const factorForIdentifier = (i: string | null) => (f: SignInFactor) => {
  return 'safeIdentifier' in f && f.safeIdentifier === i;
};

function findPasskeyStrategy(factors: SignInFactor[]): SignInFactor | null {
  if (isWebAuthnSupported()) {
    // @ts-ignore
    const passkeyFactor = factors.find(({ strategy }) => strategy === 'passkey');

    if (passkeyFactor) {
      return passkeyFactor;
    }
  }
  return null;
}

function determineStrategyWhenPasswordIsPreferred(
  factors: SignInFactor[],
  identifier: string | null,
): SignInFactor | null {
  const passkeyFactor = findPasskeyStrategy(factors);
  if (passkeyFactor) {
    return passkeyFactor;
  }
  const selected = factors.sort(passwordPrefFactorComparator)[0];
  if (selected.strategy === 'password') {
    return selected;
  }
  return factors.find(factorForIdentifier(identifier)) || selected || null;
}

function determineStrategyWhenOTPIsPreferred(factors: SignInFactor[], identifier: string | null): SignInFactor | null {
  const passkeyFactor = findPasskeyStrategy(factors);
  if (passkeyFactor) {
    return passkeyFactor;
  }
  const sortedBasedOnPrefFactor = factors.sort(otpPrefFactorComparator);
  const forIdentifier = sortedBasedOnPrefFactor.find(factorForIdentifier(identifier));
  if (forIdentifier) {
    return forIdentifier;
  }
  const firstBasedOnPref = sortedBasedOnPrefFactor[0];
  if (firstBasedOnPref.strategy === 'email_link') {
    return firstBasedOnPref;
  }
  return factors.find(factorForIdentifier(identifier)) || firstBasedOnPref || null;
}

// The algorithm can be found at
// https://www.notion.so/clerkdev/Implement-sign-in-alt-methods-e6e60ffb644645b3a0553b50556468ce
export function determineStartingSignInFactor(
  firstFactors: SignInFactor[] | null,
  identifier: string | null,
  preferredSignInStrategy: PreferredSignInStrategy,
): SignInFactor | null | undefined {
  if (!firstFactors || firstFactors.length === 0) {
    return null;
  }

  return preferredSignInStrategy === PREFERRED_SIGN_IN_STRATEGIES.Password
    ? determineStrategyWhenPasswordIsPreferred(firstFactors, identifier)
    : determineStrategyWhenOTPIsPreferred(firstFactors, identifier);
}

export function determineSalutation(signIn: Partial<SignInResource>): string {
  if (!signIn) {
    return '';
  }

  return titleize(signIn.userData?.firstName) || titleize(signIn.userData?.lastName) || signIn?.identifier || '';
}

const localStrategies: SignInStrategy[] = ['passkey', 'email_code', 'password', 'phone_code', 'email_link'];

export function factorHasLocalStrategy(factor: SignInFactor | undefined | null): boolean {
  if (!factor) {
    return false;
  }
  return localStrategies.includes(factor.strategy);
}

// The priority of second factors is: TOTP -> Phone code -> any other factor
export function determineStartingSignInSecondFactor(secondFactors: SignInFactor[] | null): SignInFactor | null {
  if (!secondFactors || secondFactors.length === 0) {
    return null;
  }

  const totpFactor = secondFactors.find(f => f.strategy === 'totp');
  if (totpFactor) {
    return totpFactor;
  }

  const phoneCodeFactor = secondFactors.find(f => f.strategy === 'phone_code');
  if (phoneCodeFactor) {
    return phoneCodeFactor;
  }

  return secondFactors[0];
}

const resetPasswordStrategies: SignInStrategy[] = ['reset_password_phone_code', 'reset_password_email_code'];
export const isResetPasswordStrategy = (strategy: SignInStrategy | string | null | undefined) =>
  !!strategy && resetPasswordStrategies.includes(strategy as SignInStrategy);

const isEmail = (str: string) => /^\S+@\S+\.\S+$/.test(str);
export function getSignUpAttributeFromIdentifier(identifier: FormControlState<'identifier'>) {
  if (identifier.type === 'tel') {
    return 'phoneNumber';
  }

  if (isEmail(identifier.value)) {
    return 'emailAddress';
  }

  return 'username';
}
