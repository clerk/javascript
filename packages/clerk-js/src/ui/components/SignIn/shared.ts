import { isClerkRuntimeError, isUserLockedError } from '@clerk/shared/error';
import { useClerk } from '@clerk/shared/react';
import { useCallback, useEffect } from 'react';

import { clerkInvalidFAPIResponse } from '../../../core/errors';
import { __internal_WebAuthnAbortService } from '../../../utils/passkeys';
import { useCoreSignIn, useSignInContext } from '../../contexts';
import { useCardState } from '../../elements';
import { useSupportEmail } from '../../hooks/useSupportEmail';
import { handleError } from '../../utils';

function useHandleAuthenticateWithPasskey(onSecondFactor: () => Promise<unknown>) {
  const card = useCardState();
  // @ts-expect-error -- private method for the time being
  const { setActive, __internal_navigateWithError } = useClerk();
  const supportEmail = useSupportEmail();
  const { afterSignInUrl } = useSignInContext();
  const { authenticateWithPasskey } = useCoreSignIn();

  useEffect(() => {
    return () => {
      __internal_WebAuthnAbortService.abort();
    };
  }, []);

  return useCallback(async (...args: Parameters<typeof authenticateWithPasskey>) => {
    try {
      const res = await authenticateWithPasskey(...args);
      switch (res.status) {
        case 'complete':
          return setActive({ session: res.createdSessionId, redirectUrl: afterSignInUrl });
        case 'needs_second_factor':
          return onSecondFactor();
        default:
          return console.error(clerkInvalidFAPIResponse(res.status, supportEmail));
      }
    } catch (err) {
      const { flow } = args[0] || {};

      if (isClerkRuntimeError(err)) {
        // In any case if the call gets aborted we should skip showing an error. This prevents updating the state of unmounted components.
        if (err.code === 'passkey_operation_aborted') {
          return;
        }
        // In case of autofill, if retrieval of credentials is cancelled by the user avoid showing errors as it results to pour UX.
        if (flow === 'autofill' && err.code === 'passkey_retrieval_cancelled') {
          return;
        }
      }

      if (isUserLockedError(err)) {
        return __internal_navigateWithError('..', err.errors[0]);
      }
      handleError(err, [], card.setError);
    }
  }, []);
}

export { useHandleAuthenticateWithPasskey };
