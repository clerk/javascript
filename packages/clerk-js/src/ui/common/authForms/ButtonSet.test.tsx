import { render, renderJSON, screen } from '@clerk/shared/testUtils';
import { titleize } from '@clerk/shared/utils/string';
import userEvent from '@testing-library/user-event';
import * as React from 'react';

import { ButtonSet } from './ButtonSet';

function MockIcon() {
  return (
    <svg>
      <circle
        cx='40'
        cy='40'
        r='24'
      />
    </svg>
  );
}

const options = [
  {
    id: 'facebook',
    name: 'Facebook',
    icon: MockIcon,
    strategy: 'oauth_facebook',
  },
  {
    id: 'google',
    name: 'Google',
    icon: MockIcon,
    strategy: 'oauth_google',
  },
];

describe('<ButtonSet/>', () => {
  it('renders the sign in button set for the provided options', () => {
    const mockHandleClick = jest.fn();
    const tree = renderJSON(
      <ButtonSet
        options={options}
        flow='sign-in'
        handleClick={mockHandleClick}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it.each(['google', 'facebook'])('renders the sign up button set and clicks on %s', async (provider: string) => {
    const providerTitle = titleize(provider);

    const mockHandleClick = jest.fn();

    render(
      <ButtonSet
        options={options}
        flow='sign-up'
        handleClick={mockHandleClick}
      />,
    );

    const regex = new RegExp(`Sign up with ${providerTitle}`, 'i');

    await userEvent.click(screen.getByRole('button', { name: regex }));

    expect(mockHandleClick.mock.calls[0][1]).toBe(`oauth_${provider}`);
  });
});
