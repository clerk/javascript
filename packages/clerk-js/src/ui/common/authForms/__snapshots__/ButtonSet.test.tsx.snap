// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonSet/> renders the sign in button set for the provided options 1`] = `
<div>
  <button
    className="button outline primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <img
      alt="Facebook"
      src="https://images.clerk.dev/static/facebook.svg"
    />
    <span>
      Sign in with Facebook
    </span>
  </button>
  <button
    className="button outline primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <img
      alt="Google"
      src="https://images.clerk.dev/static/google.svg"
    />
    <span>
      Sign in with Google
    </span>
  </button>
</div>
`;
